lighthouse-core/audits/polyfills.js

```diff
@@ -0,0 +1,207 @@
+ */
+'use strict';
+
+/** @typedef {{object: string | undefined, property: string}} Poly */
+/** @typedef {{row: number, col: number, poly: Poly}} PolyIssue */
+
+const Audit = require('./audit');
+const i18n = require('../lib/i18n/i18n.js');
+
+const UIStrings = {
+  /** Imperative title of a Lighthouse audit that tells the user to remove JavaScript that is never evaluated during page load. This is displayed in a list of audit titles that Lighthouse generates. */
+  title: 'Polyfills',
+  /** Description of a Lighthouse audit that tells the user *why* they should remove JavaScript that is never needed/evaluated by the browser. This is displayed after a user expands the section to see more. No character length limits. 'Learn More' becomes link text to additional documentation. */
+  // eslint-disable-next-line max-len
+  description: 'Polyfills enable older browsers to use new JavaScript language features. However, they aren\'t always necessary. Research what browsers you must support and consider removing polyfils for features that are well supported by them.',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+class Polyfills extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'polyfills',
+      scoreDisplayMode: Audit.SCORING_MODES.MANUAL,
+      description: str_(UIStrings.description),
+      title: str_(UIStrings.title),
+      requiredArtifacts: ['Scripts'],
+    };
+  }
+
+  /**
+     * @param {Poly[]} polys
+     * @param {string} code
+     * @return {PolyIssue[]}
+     */
+  static detectPolys(polys, code) {
+    const qt = (/** @type {string} */ token) =>
+      `['"]${token}['"]`; // don't worry about matching string delims
+    const pattern = polys.map(({object, property}) => {
+      let subpattern = '';
+
+      // String.prototype.startsWith =
+      subpattern += `${object || ''}.${property}\\s*=`;
+      subpattern += `|${object || ''}\\[${qt(property)}\\]\\s*=`;
+
+      // Object.defineProperty(String.prototype, 'startsWith'
+      subpattern += `|defineProperty\\(${object || qt('window')},\\s*${qt(property)}`;
+
+      // check for lodash. Assume that there is a lodash module for most polys.
+      // TODO this would only work if the bundle has not been stripped of module names (and replaced with an index into the bundled modules),
+      // which is unlikely in a production environment. Instead, look at source map?
+      // TODO how to detect if all of lodash is bundled (a waste)?
+      // TODO maybe just punt entirely for now?
+      // subpattern += `|lodash/${property}`;
+
+      return `(${subpattern})`;
+    }).join('|');
+
+    const polysSeen = new Set();
+    const polyMatches = [];
+    const re = new RegExp(`(^\r\n|\r|\n)|${pattern}`, 'g');
+    /** @type {RegExpExecArray | null} */
+    let result;
+    let row = 0;
+    let rowBeginsAtIndex = 0;
+    while ((result = re.exec(code)) !== null) {
```

Comment:
I'm gonna at least need a comment here :)

the `...matches` below made me think the results were already somehow all inside one regex match, but the `.exec` suggests we need to iterate through them all

the `..matches` is just the subgroups of each match?

Reply:
yeah, and only one subgroup is ever defined in each `matches`. I agree this is arcane bordering on magic. I'll comment accordingly.